### PR TITLE
Feature/search functionality

### PR DESCRIPTION
**Pull request by mgupta0421**
This PR implements the following functionalities and configurations from the frontend and backend part-
Backend-
1- The search functionality
Backend- marshalling the query send by frontend, creation of HTTP request to webAPI, retrieving the result from webAPI and sending it to frontend
2-Handling of the static fileS(example index.html) which are generated by frontend build and the backend deploy it on the server...
3- Webpack configuration for integration of backend and frontend code.

Frontend -
Feature Branch landingPage contains the following:
• Pleasing layout
• A query text field where the user can manually type in and/or three steps (query menu) to create a query. After correctly finishing one step, the box border (div) where the step is inside will become green. In the second step, if the user types in a non-numerical value then the box border will be red and an information for the user will be shown (“value should be numerical”). After completing successfully all steps, the “add to query” button will become green and usable. Clicking on that will add the created query to the query text field always on the right of the existing query. The steps will be reset.
Some more information to the steps:
o 1. Metric step has a selection menu to choose a metric. After choosing a metric, the selection menu will be invisible and the chosen metric will be separately shown to the user. The user can click on a rechoose button to choose a different metric. Then the selection menu will be visible again.
o 2. Operator and Value step has an operator drop down menu and a text field for the value. The value has to be numerical.
o 3. Logical Operator step has an optional checkbox that allows the user to add a logical NOT and four radio buttons for the logical operators.
Start search button is initially grey and not usable. If there is something inside the query textbox (condition will be changed later), the button becomes red and is usable. After clicking on the start search button, the query string will be sent to the backend via a function.
• A Result table that depicts the data from the backend. For now only the essential elements of the table are available.
Branch link- https://github.com/delphi-hub/delphi-webapp/tree/feature/landingPage

### CODE REVIEW

File: frontend/public/index.html [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  <link href="https://cdn.jsdelivr.net/npm/@mdi/font@4.x/css/materialdesignicons.min.css" rel="stylesheet">
+  <link rel="icon" type="image" :src="require('../assets/Logo.png')">
+  <title>Delphi - Tools and Facts for open-source projects</title>
+  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:100,300,400,500,700,900">
+  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@mdi/font@latest/css/materialdesignicons.min.css">
+</head>
+
+<body>
+  <noscript>
+    <strong>We're sorry but frontend doesn't work properly without JavaScript enabled. Please enable it to
```
Comment (bhermann) [inline frontend/public/index.html]: Exchange with: "This website requires JavaScript to be enabled. Please active it to continue. Thank you!"

File: frontend/src/components/landingPage/MenuStepOne.vue [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+				}
+				else {
+					select.options[i].style.display = 'none';
+				}
+			}
+		}
+	},
+	mounted () {
+		axios
+		.get('https://delphi.cs.uni-paderborn.de/api/features')
```
Comment (bhermann) [inline frontend/src/components/landingPage/MenuStepOne.vue]: Why is there a hard-coded address here?
This doesn't look right.

File: frontend/src/components/landingPage/ResultPart.vue [hunk 3]
```diff
@@ -0,0 +1,277 @@
+<template>
```
Comment (bhermann) [inline frontend/src/components/landingPage/ResultPart.vue]: Why is this page part of the pull request? 
The feature is not yet complete. Please only include completed features.

File: project/WebpackServer.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// limitations under the License.
+import java.io.File
+import java.net.InetSocketAddress
+
+import com.typesafe.config._
+import play.sbt.PlayRunHook
+
+import scala.sys.process.Process
+
+object WebpackServer {
```
Comment (bhermann) [inline project/WebpackServer.scala]: Is this still necessary?

**Comment by bhermann** [comment]
Happy New Year everyone!
I hope you all had a great start. 
For this PR I am still waiting for replies and changes to move this ahead.

**Comment by bhermann** [comment]
Just had a look in the branch. There are more files that shouldn't be there... 
Starting with `Future.successful...`. How do you even create that? Why do you commit those, and why do you push those?

**Comment by Subbup101** [comment]
> Just had a look in the branch. There are more files that shouldn't be there...
> Starting with `Future.successful...`. How do you even create that? Why do you commit those, and why do you push those?

Sorry for that, my bad. I am not sure how those files are created but I realised just now.

File: frontend/public/index.html (hunk 1 above)
Comment (bhermann) [reply frontend/public/index.html]: Whoop... Sorry... My English was broken. "active" -> "activate" 🙈

File: frontend/src/components/landingPage/MenuStepOne.vue (hunk 2 above)
Comment (syam595) [reply frontend/src/components/landingPage/MenuStepOne.vue]: This has been handled. Defined the base url for  axios api call in axios.js file.
So direct get call will not be visible or hardcoded

File: frontend/src/components/landingPage/ResultPart.vue (hunk 3 above)
Comment (Subbup101) [reply frontend/src/components/landingPage/ResultPart.vue]: It has been removed now which is not used in this version.

File: frontend/public/index.html (hunk 1 above)
Comment (Subbup101) [reply frontend/public/index.html]: Done, thanks!

File: project/WebpackServer.scala (hunk 4 above)
Comment (seena-mathew) [reply project/WebpackServer.scala]: This file is necessary to build frontend files(build.js, index.html and few images) but we had named it wrongly. We have updated the filename and have made few changes in build.sbt

File: app/controllers/Config.scala [hunk 5]
```diff
@@ -13,9 +13,10 @@
...
-package utils
+package controllers
 
-import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport
-import spray.json._
+object Config {
 
-trait JsonSupport extends SprayJsonSupport with DefaultJsonProtocol
\ No newline at end of file
+  val server="https://delphi.cs.uni-paderborn.de/api-legacy/search?pretty"
```
Comment (bhermann) [inline app/controllers/Config.scala]: Please do not hard code URLs in the code... Especially production URLs...

File: conf/application.conf [hunk 6]
```diff
@@ -40,7 +40,7 @@ akka {
 # ~~~~~
 # The secret key is used to sign Play's session cookie.
 # This must be changed for production, but we don't recommend you change it in this file.
-play.http.secret.key = "changeme"
+play.http.secret.key = "tZhx=zlpcMsSX<WNCm=u4<q`S:I^YwAsScO7]FMr0:jEwjLsq73xesM@5a7p1SFo"
```
Comment (bhermann) [inline conf/application.conf]: Do **NOT** put production secrets into the repository...

File: conf/application.conf [hunk 7]
```diff
@@ -349,11 +349,7 @@ db {
...
-app.portWebapp = 8085
-
-webapi.port = 8080
-webapi.host = "http://localhost"
-instance.registry.path = "http://localhost:8087"
-play.modules.enabled += "app.EagerLoaderModule"
\ No newline at end of file
+play.filters.hosts {
+  # Allow requests to localhost:9000.
+  allowed=["localhost:9000","127.0.0.1:9000","pg-delphi.cs.uni-paderborn.de:8080"]
```
Comment (bhermann) [inline conf/application.conf]: Again... Production configuration inside of the source code... Please remove this or replace by something configurable from the outside. (cf. DELPHI_WEBAPI_URL environment variable.)

Comment (Subbup101) [reply conf/application.conf]: I have removed it for now, we'll use it as env variable in future. Thanks!

File: conf/application.conf (hunk 6 above)
Comment (Subbup101) [reply conf/application.conf]: I have forgotten to change it back. It has been removed now. Thanks!

File: conf/application.conf [hunk 8]
```diff
@@ -273,7 +273,7 @@ play.filters {
     #permittedCrossDomainPolicies = "master-only"
 
     # The Content-Security-Policy header. If null, the header is not set.
-    contentSecurityPolicy = "default-src 'self'; script-src https://code.jquery.com https://cdnjs.cloudflare.com 'self'; style-src https://fonts.googleapis.com 'self'; font-src https://fonts.gstatic.com"
+    contentSecurityPolicy = "default-src 'self'; script-src https://code.jquery.com https://cdnjs.cloudflare.com 'self' ; style-src https://fonts.googleapis.com https://cdn.jsdelivr.net 'self' 'unsafe-inline'; font-src https://fonts.gstatic.com https://cdn.jsdelivr.net 'self' 'unsafe-inline'; connect-src https://delphi.cs.uni-paderborn.de/api-legacy/features 'self' 'unsafe-inline';"
```
Comment (bhermann) [inline conf/application.conf]: Please test if you can remove the `delphi.cs...` link here. I think you don't need that anymore.

File: frontend/src/components/landingPage/axios.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
+var axios = require("axios");
+var axiosInstance = axios.create({
+  baseURL: "https://delphi.cs.uni-paderborn.de/api-legacy/features",
```
Comment (bhermann) [inline frontend/src/components/landingPage/axios.js]: Hard-coded URL again...

Comment (bhermann) [reply frontend/src/components/landingPage/axios.js]: It seems that you query the API directly from the frontend... Please consider changing that as we have discussed this several times already.

File: frontend/src/components/landingPage/MenuStepOne.vue (hunk 2 above)
Comment (bhermann) [reply frontend/src/components/landingPage/MenuStepOne.vue]: You just moved it... This is not better. :)

File: conf/application.conf (hunk 7 above)
Comment (bhermann) [reply conf/application.conf]: I'll keep this comment open as a reminder.

File: frontend/src/components/landingPage/axios.js (hunk 9 above)
Comment (Subbup101) [reply frontend/src/components/landingPage/axios.js]: It has been taken care by calling backend API which then calls Delphi WEB-API url. Thanks!

File: conf/application.conf (hunk 8 above)
Comment (Subbup101) [reply conf/application.conf]: After removing hard-coded url in frontend, links from this CS policy has been removed. Thanks!

File: conf/application.conf (hunk 7 above)
Comment (Subbup101) [reply conf/application.conf]: It has been updated with an env variable. Please check whether I did it right or not. Thanks!

File: app/controllers/Config.scala (hunk 5 above)
Comment (Subbup101) [reply app/controllers/Config.scala]: This also has been updated with env variables. Please verify. Thanks!

File: app/controllers/HomeController.scala [hunk 10]
```diff
@@ -15,47 +15,120 @@
...
-      val result = BlockingHttpClient.executeRequest(request)
-      result match {
-        case Success(response) => Future.successful(Ok(views.html.index(response, query, false)))
-        case Failure(_) => Future.successful(Ok(views.html.index("ERROR: Failed to reach server at " + request.uri, query, true)))
+
+      implicit val system = ActorSystem()
+      implicit val ec = system.dispatcher
+      implicit val materializer = ActorMaterializer()
+
+      val featuresUri = sys.env.getOrElse("DELPHI_WEBAPI_URL","https://delphi.cs.uni-paderborn.de/api-legacy") + "/features"
```
Comment (bhermann) [inline app/controllers/HomeController.scala]: Here you see another usage.

File: app/utils/Configuration.scala [hunk 11]
```diff
@@ -28,7 +28,8 @@ class Configuration(val bindPort: Int = ConfigFactory.load().getInt("app.portWeb
   val instanceName = "WebAppInstance"
   val instanceRegistryUri : String = sys.env.getOrElse("DELPHI_IR_URI", ConfigFactory.load().getString("instance.registry.path"))
 
-  lazy val webApiUri:String =  webApiInstance.host + ":" + webApiInstance.portNumber
+  lazy val webApiUri:String =  webApiInstance.host
```
Comment (bhermann) [inline app/utils/Configuration.scala]: By the way... Here was the original base URL retrieved through the instanceregistry or the default configuration.

File: app/controllers/Config.scala (hunk 5 above)
Comment (bhermann) [reply app/controllers/Config.scala]: This is resolved, however there are new comments on this issue below.

File: conf/application.conf (hunk 7 above)
Comment (bhermann) [reply conf/application.conf]: It's better now. Thanks.

File: conf/application.conf (hunk 8 above)
Comment (bhermann) [reply conf/application.conf]: Great! Thanks!

File: frontend/src/components/landingPage/axios.js (hunk 9 above)
Comment (bhermann) [reply frontend/src/components/landingPage/axios.js]: Much better now... Thank you!

**Comment by bhermann** [comment]
I'll do a runtime test over the weekend. :)

File: app/controllers/Config.scala [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package controllers
+
+object Config {
+
+  val server = sys.env.getOrElse("DELPHI_WEBAPI_URL","https://delphi.cs.uni-paderborn.de/api-legacy") + "/search?pretty"
```
[the remark this answers is not in the document]
Comment (Shashank-KanakapuraSrivatsa) [reply app/controllers/Config.scala]: This has now been addressed through CommonHelper.scala

File: app/controllers/HomeController.scala (hunk 10 above)
Comment (Shashank-KanakapuraSrivatsa) [reply app/controllers/HomeController.scala]: This has now been addressed through CommonHelper.scala

File: app/utils/Configuration.scala (hunk 11 above)
Comment (Shashank-KanakapuraSrivatsa) [reply app/utils/Configuration.scala]: Issues have been addressed

Comment (bhermann) [reply app/utils/Configuration.scala]: kind of... :)

**Pull request merged**